File: go/uuid/uuid.go
// Package uuid implements RFC 9562 version 4 (random) and version 7
// (time-ordered) UUIDs. The Go standard library has no UUID type.
//
// A UUIDv7 is a 48-bit big-endian millisecond Unix timestamp, the version
// nibble, 12 random bits, the variant bits and 62 more random bits. A UUIDv4
// is 122 random bits plus the version and variant bits.
package uuid

import (
	"crypto/rand"
	"errors"
	"sync"
	"time"
)

// UUID is the 16-byte binary form, in the same byte order as the text form,
// so bytes.Compare order matches string order.
type UUID [16]byte

const (
	// EncodedSize is the length of the canonical string form.
	EncodedSize = 36
	// MaxTime is the largest UUIDv7 timestamp in milliseconds.
	MaxTime uint64 = 1<<48 - 1

	randBMax  = 1<<62 - 1
	hexDigits = "0123456789abcdef"
)

var (
	ErrInvalidFormat = errors.New("uuid: must be 36 characters in 8-4-4-4-12 hex form")
	ErrTimeRange     = errors.New("uuid: timestamp exceeds 48 bits")
	ErrMonotonic     = errors.New("uuid: monotonic random component overflow")
	ErrNotV7         = errors.New("uuid: not a version 7 UUID")
)

func nowMs() uint64 { return uint64(time.Now().UnixMilli()) }

func fillRandom(b []byte) {
	if _, err := rand.Read(b); err != nil {
		panic("uuid: crypto/rand failed: " + err.Error())
	}
}

func (u *UUID) setVersion(v byte) {
	u[6] = u[6]&0x0F | v<<4
	u[8] = u[8]&0x3F | 0x80
}

// NewV4 returns a random (version 4) UUID.
func NewV4() UUID {
	var u UUID
	fillRandom(u[:])
	u.setVersion(4)
	return u
}

// V4FromRandom sets the version and variant bits of 16 random bytes.
func V4FromRandom(random [16]byte) UUID {
	u := UUID(random)
	u.setVersion(4)
	return u
}

// NewV7 returns a time-ordered (version 7) UUID for the current time.
func NewV7() UUID {
	var rnd [10]byte
	fillRandom(rnd[:])
	u, err := V7FromParts(nowMs(), rnd)
	if err != nil {
		panic(err)
	}
	return u
}

// V7FromParts builds a UUIDv7 from a millisecond timestamp and 10 random
// bytes; the version and variant bits overwrite 6 of the random bits.
func V7FromParts(ms uint64, random [10]byte) (UUID, error) {
	var u UUID
	if ms > MaxTime {
		return u, ErrTimeRange
	}
	for i := 5; i >= 0; i-- {
		u[i] = byte(ms)
		ms >>= 8
	}
	copy(u[6:], random[:])
	u.setVersion(7)
	return u, nil
}

// Version returns the version nibble (4 for NewV4, 7 for NewV7).
func (u UUID) Version() int { return int(u[6] >> 4) }

// Time returns the Unix timestamp in milliseconds of a version 7 UUID.
func (u UUID) Time() (uint64, error) {
	if u.Version() != 7 {
		return 0, ErrNotV7
	}
	var ms uint64
	for i := 0; i < 6; i++ {
		ms = ms<<8 | uint64(u[i])
	}
	return ms, nil
}

// String returns the canonical lowercase 36-character form.
func (u UUID) String() string {
	var buf [EncodedSize]byte
	u.encode(&buf)
	return string(buf[:])
}

// AppendText appends the canonical encoding to b.
func (u UUID) AppendText(b []byte) ([]byte, error) {
	var buf [EncodedSize]byte
	u.encode(&buf)
	return append(b, buf[:]...), nil
}

// MarshalText implements encoding.TextMarshaler.
func (u UUID) MarshalText() ([]byte, error) {
	var buf [EncodedSize]byte
	u.encode(&buf)
	return buf[:], nil
}

// UnmarshalText implements encoding.TextUnmarshaler.
func (u *UUID) UnmarshalText(b []byte) error {
	v, err := parse(b)
	if err != nil {
		return err
	}
	*u = v
	return nil
}

func (u UUID) encode(dst *[EncodedSize]byte) {
	o := 0
	for i, b := range u {
		if i == 4 || i == 6 || i == 8 || i == 10 {
			dst[o] = '-'
			o++
		}
		dst[o], dst[o+1] = hexDigits[b>>4], hexDigits[b&15]
		o += 2
	}
}

// Parse decodes the case-insensitive 8-4-4-4-12 form. Braces, the urn:uuid:
// prefix and the 32-digit form without hyphens are rejected.
func Parse(s string) (UUID, error) { return parse([]byte(s)) }

func parse(s []byte) (UUID, error) {
	var u UUID
	if len(s) != EncodedSize {
		return u, ErrInvalidFormat
	}
	t := 0
	for i := range u {
		if t == 8 || t == 13 || t == 18 || t == 23 {
			if s[t] != '-' {
				return UUID{}, ErrInvalidFormat
			}
			t++
		}
		hi, lo := hexValue(s[t]), hexValue(s[t+1])
		if hi < 0 || lo < 0 {
			return UUID{}, ErrInvalidFormat
		}
		u[i] = byte(hi<<4 | lo)
		t += 2
	}
	return u, nil
}

func hexValue(c byte) int {
	switch {
	case c >= '0' && c <= '9':
		return int(c - '0')
	case c|0x20 >= 'a' && c|0x20 <= 'f':
		return int(c|0x20-'a') + 10
	}
	return -1
}

// Monotonic generates strictly increasing UUIDv7s and is safe for concurrent
// use. Within the same millisecond (or if the clock moves backwards) the
// previous 74 random bits are incremented by one instead of drawing new ones.
type Monotonic struct {
	mu     sync.Mutex
	lastMs uint64
	last   UUID
	primed bool
}

// NewMonotonic returns a ready-to-use monotonic generator.
func NewMonotonic() *Monotonic { return &Monotonic{} }

// Next returns the next UUIDv7, or ErrMonotonic if 2^74 IDs were requested in
// a single millisecond.
func (m *Monotonic) Next() (UUID, error) {
	return m.next(nowMs(), fillRandom)
}

func (m *Monotonic) next(now uint64, fill func([]byte)) (UUID, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	if m.primed && now <= m.lastMs {
		next, ok := increment(m.last)
		if !ok {
			return UUID{}, ErrMonotonic
		}
		m.last = next
		return next, nil
	}
	if now > MaxTime {
		return UUID{}, ErrTimeRange
	}
	var rnd [10]byte
	fill(rnd[:])
	m.last, _ = V7FromParts(now, rnd)
	m.lastMs = now
	m.primed = true
	return m.last, nil
}

// increment adds one to the 74 random bits (rand_a<<62 | rand_b), skipping
// the version and variant bits.
func increment(u UUID) (UUID, bool) {
	randA := uint64(u[6]&0x0F)<<8 | uint64(u[7])
	randB := uint64(u[8] & 0x3F)
	for i := 9; i < 16; i++ {
		randB = randB<<8 | uint64(u[i])
	}
	switch {
	case randB < randBMax:
		randB++
	case randA < 0xFFF:
		randA, randB = randA+1, 0
	default:
		return u, false
	}
	u[6], u[7] = 0x70|byte(randA>>8), byte(randA)
	for i := 15; i >= 9; i-- {
		u[i] = byte(randB)
		randB >>= 8
	}
	u[8] = 0x80 | byte(randB)
	return u, true
}

File: rust/src/uuid.rs
//! RFC 9562 UUIDs: version 4 (122 random bits) and version 7 (48-bit
//! millisecond timestamp + 74 random bits). The standard library has no UUID
//! type, so this is a small dependency-free one.

use crate::{now_ms, rng, Error};
use std::fmt;
use std::str::FromStr;
use std::sync::Mutex;

/// Largest UUIDv7 timestamp in milliseconds.
pub const MAX_TIME: u64 = (1 << 48) - 1;
/// Length of the canonical string form.
pub const ENCODED_LEN: usize = 36;

const HEX: &[u8; 16] = b"0123456789abcdef";
const RAND_B_MASK: u128 = (1 << 62) - 1;
const RAND_MAX: u128 = (1 << 74) - 1;

/// A 128-bit UUID. Ordering matches both byte order and string order.
#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
pub struct Uuid(pub u128);

fn with_version(v: u128, version: u8) -> Uuid {
    let v = (v & !(0xF << 76)) | ((version as u128) << 76);
    Uuid((v & !(0b11 << 62)) | (0b10 << 62))
}

impl Uuid {
    /// Random (version 4) UUID from OS randomness.
    pub fn new_v4() -> Uuid {
        let mut r = [0u8; 16];
        rng::fill(&mut r);
        Uuid::v4_from_bytes(r)
    }

    /// Sets the version and variant bits of 16 random bytes.
    pub fn v4_from_bytes(random: [u8; 16]) -> Uuid {
        with_version(u128::from_be_bytes(random), 4)
    }

    /// Time-ordered (version 7) UUID for the current time.
    pub fn new_v7() -> Result<Uuid, Error> {
        let mut r = [0u8; 10];
        rng::fill(&mut r);
        Uuid::v7_from_parts(now_ms(), r)
    }

    /// Builds a UUIDv7; the version and variant bits overwrite 6 of the random bits.
    pub fn v7_from_parts(timestamp_ms: u64, random: [u8; 10]) -> Result<Uuid, Error> {
        if timestamp_ms > MAX_TIME {
            return Err(Error::TimeRange);
        }
        let mut b = [0u8; 16];
        b[..6].copy_from_slice(&timestamp_ms.to_be_bytes()[2..]);
        b[6..].copy_from_slice(&random);
        Ok(with_version(u128::from_be_bytes(b), 7))
    }

    /// The version nibble (4 or 7 for UUIDs made here).
    pub fn version(self) -> u8 {
        (self.0 >> 76) as u8 & 0xF
    }

    /// Unix timestamp in milliseconds of a version 7 UUID.
    pub fn timestamp_ms(self) -> Result<u64, Error> {
        if self.version() != 7 {
            return Err(Error::NotUuidV7);
        }
        Ok((self.0 >> 80) as u64)
    }

    pub fn to_bytes(self) -> [u8; 16] {
        self.0.to_be_bytes()
    }

    pub fn from_bytes(bytes: [u8; 16]) -> Uuid {
        Uuid(u128::from_be_bytes(bytes))
    }

    /// Encode the lowercase 8-4-4-4-12 form into a stack buffer without allocating.
    pub fn encode(self, out: &mut [u8; ENCODED_LEN]) {
        let mut o = 0;
        for (i, b) in self.to_bytes().into_iter().enumerate() {
            if matches!(i, 4 | 6 | 8 | 10) {
                out[o] = b'-';
                o += 1;
            }
            out[o] = HEX[(b >> 4) as usize];
            out[o + 1] = HEX[(b & 15) as usize];
            o += 2;
        }
    }

    /// Case-insensitive 8-4-4-4-12 form only: braces, `urn:uuid:` and the
    /// 32-digit form without hyphens are rejected.
    pub fn parse(s: &str) -> Result<Uuid, Error> {
        let b = s.as_bytes();
        if b.len() != ENCODED_LEN {
            return Err(Error::InvalidUuid);
        }
        let mut v: u128 = 0;
        for (i, &c) in b.iter().enumerate() {
            if matches!(i, 8 | 13 | 18 | 23) {
                if c != b'-' {
                    return Err(Error::InvalidUuid);
                }
                continue;
            }
            let d = (c as char).to_digit(16).ok_or(Error::InvalidUuid)?;
            v = (v << 4) | d as u128;
        }
        Ok(Uuid(v))
    }

    fn rand74(self) -> u128 {
        ((self.0 >> 64) & 0xFFF) << 62 | (self.0 & RAND_B_MASK)
    }

    fn with_rand74(self, r: u128) -> Uuid {
        let time = self.0 >> 80 << 80;
        Uuid(time | 7 << 76 | (r >> 62) << 64 | 0b10 << 62 | (r & RAND_B_MASK))
    }
}

impl fmt::Display for Uuid {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut buf = [0u8; ENCODED_LEN];
        self.encode(&mut buf);
        // SAFETY: every byte is an ASCII hex digit or '-'.
        f.write_str(unsafe { std::str::from_utf8_unchecked(&buf) })
    }
}

impl fmt::Debug for Uuid {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Uuid({self})")
    }
}

impl FromStr for Uuid {
    type Err = Error;
    fn from_str(s: &str) -> Result<Uuid, Error> {
        Uuid::parse(s)
    }
}

/// Thread-safe generator of strictly increasing UUIDv7s.
///
/// Within the same millisecond (or if the clock moves backwards) the previous
/// 74 random bits are incremented instead of drawing fresh randomness.
#[derive(Default)]
pub struct MonotonicV7Generator {
    last: Mutex<Option<Uuid>>,
}

impl MonotonicV7Generator {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn next(&self) -> Result<Uuid, Error> {
        self.next_at(now_ms(), rng::fill)
    }

    pub(crate) fn next_at(&self, now: u64, fill: impl FnOnce(&mut [u8])) -> Result<Uuid, Error> {
        let mut last = self.last.lock().unwrap_or_else(|e| e.into_inner());
        if let Some(prev) = *last {
            if now <= (prev.0 >> 80) as u64 {
                let r = prev.rand74();
                if r == RAND_MAX {
                    return Err(Error::MonotonicOverflow);
                }
                let next = prev.with_rand74(r + 1);
                *last = Some(next);
                return Ok(next);
            }
        }
        let mut r = [0u8; 10];
        if now <= MAX_TIME {
            fill(&mut r);
        }
        let next = Uuid::v7_from_parts(now, r)?;
        *last = Some(next);
        Ok(next)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_vectors;

    #[test]
    fn monotonic_vectors() {
        let rows = test_vectors("uuid7_monotonic.txt");
        assert!(rows.len() > 10);
        let mut g = MonotonicV7Generator::new();
        for r in &rows {
            if r[0] == "reset" {
                g = MonotonicV7Generator::new();
                continue;
            }
            let now: u64 = r[1].parse().unwrap();
            let mut drew = false;
            let got = g.next_at(now, |buf| {
                drew = true;
                if r[2] != "-" {
                    let rnd = u128::from_str_radix(&r[2], 16).unwrap();
                    buf.copy_from_slice(&rnd.to_be_bytes()[6..]);
                }
            });
            assert!(!(r[2] == "-" && drew), "{r:?}: drew randomness");
            match (r[3].as_str(), got) {
                ("error", Ok(u)) => panic!("{r:?}: got {u}, want an error"),
                ("error", Err(_)) => {}
                (want, got) => assert_eq!(got.map(|u| u.to_string()), Ok(want.to_string()), "{r:?}"),
            }
        }
    }
}
